### ISSUE TITLE
[python-package] Why is using CUDA different from CPU results

### ISSUE THREAD

**Issue by guoqianghao9** [microsoft/LightGBM#6055]
os: ubuntu 22.03LTS
lightgbm version = 4.0.0 install version : 

```shell
cmake -D CUDA_TOOLKIT_ROOT_DIR=/usr/local/cuda -DUSE_CUDA=1 ..
pip install --no-binary lightgbm --config-settings=cmake.define.USE_CUDA=ON lightgbm
```

single GPU : nvidia 3090 
cuda : 11.8
cudnn: 8.6.0.163

```python
import lightgbm as lgb
import numpy as np
from sklearn.model_selection import train_test_split

np.random.seed(42)
n_samples = 500*10000
n_features = 51
X = np.random.rand(n_samples,n_features)
y = np.random.rand(n_samples)
X = np.round(X,8)
y = np.round(y,8)
X_train , X_test ,y_train , y_test = train_test_split(X,y,test_size=0.2,random_state=42)
new_lgb_train = lgb.Dataset(X_train, y_train)

cpu_params  = {'objective':'regression',
    'boosting_type':'dart', 
     'colsample_bytree': 0.7, 
    'learning_rate': 0.01,
     'max_depth' :7, 
     'subsample' : 0.7,
        'n_jobs':32,
    'num_leaves': 63,
    'verbose': 1,
    'device':'cpu'
}

cuda_params  = {'objective':'regression',
    'boosting_type':'dart', 
     'colsample_bytree': 0.7, 
    'learning_rate': 0.01,
     'max_depth' :7, 
     'subsample' : 0.7,
        'n_jobs':32,
    'num_leaves': 63,
    'verbose': 1,
    'device':'cuda'
}
gbm_cuda = lgb.train(cuda_params,
                new_lgb_train,
                num_boost_round=100)

gbm_cpu = lgb.train(cpu_params,
                new_lgb_train,
                num_boost_round=100)

gbm_cpu.predict(X_test)

# out : array([0.31292154, 0.31217245, 0.31260493, ..., 0.312324  , 0.31271839,
#       0.31371233])

gbm_cuda.predict(X_test)
# out: array([0.31276433, 0.31235974, 0.31262128, ..., 0.31243856, 0.3127279 ,
#       0.31381523])
```

BTW:when i run `gbm_cuda`, I don't know if this will affect the calculation results.

jupyter notebook prompt :

> [LightGBM] [Warning] Using sparse features with CUDA is currently not supported.

**Comment by jameslamb** [comment]
I've updated this report to make the difference between your own words, code, and log output from code clearer. For the future, please see https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax to learn how to format text on GitHub.

**Comment by jameslamb** [comment]
So first.... running the following

```shell
pip install --no-binary lightgbm --config-settings=cmake.define.USE_CUDA=ON lightgbm
```

Will download the source distribution of `lightgbm` from PyPI and then compile it. If you're going to install `lightgbm` from PyPI that way, that makes this command unnecessary:

```shell
cmake -D CUDA_TOOLKIT_ROOT_DIR=/usr/local/cuda -DUSE_CUDA=1 ..
```

**Comment by jameslamb** [comment]
Second... I suspect that numerical differences of the form shown in your results could happen **even for 2 consecutive runs on CPU**. That's because there is some inherent lack of determinism in each run, especially since you're explicitly asking LightGBM to randomly sample rows with `"subsample": 0.7`.

Please see all of the advice in https://github.com/microsoft/LightGBM/issues/5934#issuecomment-1596208059 and https://github.com/microsoft/LightGBM/issues/5887#issuecomment-1545935847 for how to reduce the differences between different model-training runs.

**Comment by guoqianghao9** [comment]
> 
Is there any difference between the two approaches (pip and cmake)?
I started using
```
cmake -D CUDA_TOOLKIT_ROOT_DIR=/usr/local/cuda -DUSE_CUDA=1 ..
cmake -j8
```

It generates a lot of files for me in the build directory, I don't know how to use it, I also use make insall, I failed to import lightgbm

```
(base) ubuntu@ubuntu-MS-Challenger-B760M:~/下载/LightGBM$ ls
build                  build-python.sh  cmake           CODE_OF_CONDUCT.md  docker  examples       helpers  lib_lightgbm.so  lightgbm-python  python-package  R-package    src   test.ipynb  VERSION.txt
build-cran-package.sh  build_r.R        CMakeLists.txt  CONTRIBUTING.md     docs    external_libs  include  LICENSE          pmml             README.md       SECURITY.md  swig  tests       windows
(base) ubuntu@ubuntu-MS-Challenger-B760M:~/下载/LightGBM$ cd build/
(base) ubuntu@ubuntu-MS-Challenger-B760M:~/下载/LightGBM/build$ ls
CMakeCache.txt  CMakeFiles  cmake_install.cmake  install_manifest.txt  Makefile
```

**Comment by jameslamb** [comment]
> Is there any difference between the two approaches (pip and cmake)?

Yes.`pip install` downloads a release of the Python package from PyPI. `pip install --no-binary lightgbm` downloads a *source* distribution, which means source code that then has to be compiled on your machine. That compilation will use `CMake` to build `lib_lightgbm.so` and install it alongside Python code that can use it.

By "CMake" I think you mean "cloning the repo from GitHub and manually invoking `CMake` commands". That approach **does not install the Python package**... it is used to build `lib_lightgbm.so`, and then you have to do other work to build and install a Python package referencing that `lib_lightgbm.so`.

To summarize:

```shell
# download source distribution from PyPI, compile it with CUDA support
pip install \
    --no-binary lightgbm \
    --config-settings=cmake.define.USE_CUDA=ON \
    lightgbm

# build and install CUDA version of Python package, from source on GitHub
git clone --recursive https://github.com/microsoft/LightGBM.git
cd ./LightGBM
sh ./build-python.sh install --cuda
```

Reasons you might choose cloning the repo and building manually using `sh build-python.sh` instead:

* you want to use a version of the code that wasn't formally released
    - e.g., the latest version on the `master` branch here
    - e.g., your own fork with some changes on it
* you want to further customize the build, e.g. provide a different version of CUDA libraries, use a different compiler, compile a DEBUG build with more logging, etc.

If those don't apply to you, just use PyPI

```shell
pip install \
    --no-binary lightgbm \
    --config-settings=cmake.define.USE_CUDA=ON \
    lightgbm
```

That also has some other nice benefits like:

* don't need to separately install `CMake` or GNU Make (the build backend, `scikit-build-core`, can do that automatically)
* no need to clone the repo and clutter your environment with other files

**Comment by guoqianghao9** [comment]
> Second... I suspect that numerical differences of the form shown in your results could happen **even for 2 consecutive runs on CPU**. That's because there is some inherent lack of determinism in each run, especially since you're explicitly asking LightGBM to randomly sample rows with `"subsample": 0.7`.
> 
> Please see all of the advice in [#5934 (comment)](https://github.com/microsoft/LightGBM/issues/5934#issuecomment-1596208059) and [#5887 (comment)](https://github.com/microsoft/LightGBM/issues/5887#issuecomment-1545935847) for how to reduce the differences between different model-training runs.

when i 2 consecutive runs on CPU ,get same result 
```
cpu_params  = {'objective':'regression',
    'boosting_type':'dart', 
     'colsample_bytree': 0.7, 
    'learning_rate': 0.01,
     'max_depth' :7, 
     'subsample' : 0.7,
        'n_jobs':32,
    'num_leaves': 63,
    'verbose': 1,
    'device':'cpu'
}
```
[http://www.guoerdan.com/mhxy/lightgbm_error1.png](url)

**Comment by jameslamb** [comment]
> when i 2 consecutive runs on CPU ,get same result

That's possible. I didn't say you were *guaranteed* to get different results on repeated runs.

Some causes of non-determinism come from e.g. race conditions in multi-threaded code, which might not always vary from run-to-run.

But either way ... those `params` don't follow the advice I gave in the linked issues. For example, I expected to see `seed` and `deterministic` passed in parameters.

Please try re-running on CPU and with the CUDA version using *identical* code (other than parameter `device`) and specifying those parameters recommended in the links I provided, and report the results just as you did in your original post.

If the differences are still large (more than could be explained by small numeric-precision differences), then we'll consider investigating further.

> [http://www.guoerdan.com/mhxy/lightgbm_error1.png](https://github.com/microsoft/LightGBM/issues/url)

Not sure what this website is, but I won't click this link. Please post logs and error messages in plaintext, not as links to screenshots hosted on external sites. See this conversation from Stack Overflow to understand why you shouldn't post images of logs/screenshots when aasking for code help on the internet: https://meta.stackoverflow.com/questions/285551/why-should-i-not-upload-images-of-code-data-errors.

**Comment by guoqianghao9** [comment]
> > when i 2 consecutive runs on CPU ,get same result
> 
> That's possible. I didn't say you were _guaranteed_ to get different results on repeated runs.
> 
> Some causes of non-determinism come from e.g. race conditions in multi-threaded code, which might not always vary from run-to-run.
> 
> But either way ... those `params` don't follow the advice I gave in the linked issues. For example, I expected to see `seed` and `deterministic` passed in parameters.
> 
> Please try re-running on CPU and with the CUDA version using _identical_ code (other than parameter `device`) and specifying those parameters recommended in the links I provided, and report the results just as you did in your original post.
> 
> If the differences are still large (more than could be explained by small numeric-precision differences), then we'll consider investigating further.
> 
> > [http://www.guoerdan.com/mhxy/lightgbm_error1.png](https://github.com/microsoft/LightGBM/issues/url)
> 
> Not sure what this website is, but I won't click this link. Please post logs and error messages in plaintext, not as links to screenshots hosted on external sites. See this conversation from Stack Overflow to understand why you shouldn't post images of logs/screenshots when aasking for code help on the internet: https://meta.stackoverflow.com/questions/285551/why-should-i-not-upload-images-of-code-data-errors.

sorry...
now, i change those params,add     'deterministic','force_row_wise','seed'.
`cpu_params  = {'objective':'regression',
    'boosting_type':'dart', 
     'colsample_bytree': 0.7, 
    'learning_rate': 0.01,
     'max_depth' :7, 
     'subsample' : 1,
        'n_jobs':32,
    'num_leaves': 63,
    'verbose': 1,
    'device':'cpu',
    'deterministic':True,
    'force_row_wise':True,
    'seed':42
}`

`cuda_params  = {'objective':'regression',
    'boosting_type':'dart', 
     'colsample_bytree': 0.7, 
    'learning_rate': 0.01,
     'max_depth' :7, 
     'subsample' : 1,
        'n_jobs':32,
    'num_leaves': 63,
    'verbose': 1,
    'device':'cuda',
    'deterministic':True,
    'force_row_wise':True,
    'seed':42
}`
`gbm_cpu = lgb.train(cpu_params,
                new_lgb_train,
                num_boost_round=100)

gbm_cpu.predict(X_test)`
out:[LightGBM] [Warning] Although "deterministic" is set, the results ran by GPU may be non-deterministic.
array([0.27401207, 0.27373941, 0.27396187, ..., 0.27375929, 0.27410678,
       0.27273749])
`gbm_cuda = lgb.train(cuda_params,
                new_lgb_train,
                num_boost_round=100)
gbm_cuda.predict(X_test)`
out:[LightGBM] [Warning] Although "deterministic" is set, the results ran by GPU may be non-deterministic.
[LightGBM] [Warning] Using sparse features with CUDA is currently not supported.
[LightGBM] [Warning] Although "deterministic" is set, the results ran by GPU may be non-deterministic.
array([0.27407371, 0.2738107 , 0.27388661, ..., 0.27382703, 0.27398476,
       0.27403615])

**Comment by jameslamb** [comment]
Thanks for that! What happens if you also set `colsample_by_tree: 1.0`? Maybe there is an issue with how that works on CPU vs. CUDA.

Otherwise... it appears to me that the numerical differences in the results are not very large:

```text
array([0.27401207, 0.27373941, 0.27396187, ..., 0.27375929, 0.27410678, 0.27273749])
array([0.27407371, 0.2738107 , 0.27388661, ..., 0.27382703, 0.27398476, 0.27403615])
```

At least, small enough that the difference may be attributable to just numerical precision issues (which may be amplified by the higher level of parallelism involved in using CUDA, which is what that warning message bout the use of `deterministic` is hinting at).

-----

By the way, you could make this a bit easier to read and reduce the risk of errors by not duplicating the parameter definitions. Like this:

```python
params  = {
    "objective": "regression",
    "boosting_type": "dart",
    "colsample_bytree": 0.7,
    "learning_rate": 0.01,
    "max_depth": 7,
    "subsample": 1,
    "n_jobs": 32,
    "num_leaves": 63,
    "verbose": 1,
    "deterministic": True,
    "force_row_wise": True,
    "seed": 42
}

gbm_cpu = lgb.train(
    params={**params, "device": "cpu"},
    train_set=new_lgb_train,
    num_boost_round=100
)
gbm_cpu.predict(X_test)

gbm_cuda = lgb.train(
    params={**params, "device": "cuda"},
    train_set=new_lgb_train,
    num_boost_round=100
)
gbm_cuda.predict(X_test)
```

I've reformatted your message. As I said previously, please spend some time reading https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax. It will help you to write text on GitHub in a way that makes the different between your own word, code, and output from code clearer. Click "edit" on your most recent post to see the raw text I typed.

**Comment by guoqianghao9** [comment]
> colsample_by_tree

add colsample_bytree
```
params  = {
    "objective": "regression",
    "boosting_type": "dart",
    "colsample_bytree": 0.7,
    "learning_rate": 0.01,
    "max_depth": 7,
    "subsample": 1,
    "n_jobs": 32,
    "num_leaves": 63,
    "verbose": 1,
    "deterministic": True,
    "force_row_wise": True,
    "seed": 42,
    "colsample_bytree":1
}

gbm_cpu = lgb.train(
    params={**params, "device": "cpu"},
    train_set=new_lgb_train,
    num_boost_round=100
)
gbm_cpu.predict(X_test)

gbm_cuda = lgb.train(
    params={**params, "device": "cuda"},
    train_set=new_lgb_train,
    num_boost_round=100
)
gbm_cuda.predict(X_test)
```

cpu:
```
array([0.27394593, 0.27356086, 0.27396161, ..., 0.27371579, 0.27417218,
       0.2713827 ])
```
cuda:
```
array([0.27405374, 0.27364199, 0.27392612, ..., 0.27378094, 0.27403964,
       0.27407788])
```

**Comment by jameslamb** [comment]
You specified `colsample_bytree` twice. Please remove the one where you're setting it to `0.7`.

**Comment by guoqianghao9** [comment]
> You specified `colsample_bytree` twice. Please remove the one where you're setting it to `0.7`.

ok, i remove `0.7,` The result is the same
```
array([0.27394593, 0.27356086, 0.27396161, ..., 0.27371579, 0.27417218,
       0.2713827 ])
array([0.27405374, 0.27364199, 0.27392612, ..., 0.27378094, 0.27403964,
       0.27407788])
```

**Comment by guoqianghao9** [comment]
> You specified `colsample_bytree` twice. Please remove the one where you're setting it to `0.7`.

run
`gbm_cuda.predict(X_test)`
out:
`[LightGBM] [Warning] Using sparse features with CUDA is currently not supported.`
Does it affect the result?

**Comment by jameslamb** [comment]
Ok, thank you! I can't think of any other ways to reduce the difference between the two. Is that magnitude of difference meaningful for the problem you're working?

@shiyu1994 could you explain some specific ways that the CUDA implementation might produce a different trained model for DART boosting, `"objective": "regression"`, and all continuous features? For example, different data types used in GPU memory vs. CPU memory?

I'm not sure about anything more specific than "numerical precision differences caused by different degrees of parallelism".

**Comment by guoqianghao9** [comment]
> Ok, thank you! I can't think of any other ways to reduce the difference between the two. Is that magnitude of difference meaningful for the problem you're working?
> 
> @shiyu1994 could you explain some specific ways that the CUDA implementation might produce a different trained model for DART boosting, `"objective": "regression"`, and all continuous features? For example, different data types used in GPU memory vs. CPU memory?
> 
> I'm not sure about anything more specific than "numerical precision differences caused by different degrees of parallelism".

yes,  it will affect the rank result

```
import pandas as pd
a = gbm_cpu.predict(X_test)
b = gbm_cuda.predict(X_test)
pd.Series(a).rank()
pd.Series(b).rank()
```
**a:**
```
0         516649.0
1          96918.5
2         540147.0
3         902026.0
4         618501.0
            ...   
999995    353002.0
999996    178821.5
999997    199358.5
999998    769974.5
999999      2653.0
Length: 1000000, dtype: float64
```
**b:**
```
0         687194.0
1         101685.0
2         497810.0
3         921862.5
4         351879.0
            ...   
999995    696708.0
999996    140935.0
999997    232406.5
999998    656079.0
999999    739466.0
Length: 1000000, dtype: float64
```

**Comment by guolinke** [comment]
I think this result is expected. Theoretically, it is hard to ensure deterministic results for floating-point operations, the error accumulation cause different results. If you use the same hardware + the same software environment + the same number of threads, the results could be deterministic.

**Comment by jameslamb** [comment]
> I think this result is expected.

Thank you @guolinke. This is what I suspected, but I wasn't confident.

"error accumulation in floating-point operations" is exactly what I meant by "numerical precision differences caused by different degrees of parallelism". Thank you for confirming that!

@guoqianghao9 given that, I think this issue should be closed.

Use your evaluation framework (e.g. computing metrics on a holdout set, online evaluation if you're deploying a model as an application) to determine if the model produced (either on CPU or on CUDA) is performing well enough to meet your needs.

If numerical precision issues of the form reported here are impactful enough to materially change those evaluation results, consider other approaches to avoid overfitting, like those mentioned at https://lightgbm.readthedocs.io/en/latest/Parameters-Tuning.html#deal-with-over-fitting.

**Comment by guoqianghao9** [comment]
Thanks
